src/battery-check/battery-check.c

```diff
@@ -0,0 +1,141 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "battery-util.h"
+#include "build.h"
+#include "constants.h"
+#include "errno-util.h"
+#include "glyph-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "socket-util.h"
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
+               "Checks battery level to see whether there's enough charge.\n\n"
+               "   -h --help            Show this help\n"
+               "      --version         Show package version\n",
+               program_invocation_short_name);
+}
+
+static void battery_check_send_plymouth_message(char *message) {
+        assert(message);
+
+        ssize_t w;
+        static const union sockaddr_union sa = PLYMOUTH_SOCKET;
+        _cleanup_close_ int fd = -EBADF;
+        _cleanup_free_ char *plymouth_message = NULL, *plymouth_shutdown = NULL;
+
+        if (asprintf(&plymouth_message, "M\002%c%s", (int) strlen(message), message) < 0) {
+                log_oom();
+                return;
+        }
+        if (asprintf(&plymouth_shutdown, "C\002%c%s", (int) strlen("shutdown"), "shutdown") < 0) {
```

Comment:
why not do this in a single string buffer? i.e. the two strings are going to be written in one go anyway, hence you might as well format them as one.

Reply:
hence:

```c
c = asprintf(&plymouth_message, 
             "M\x02%c%s"
             "C\x02\x08shutdown",
             strlen(message), message);
```

Reply:
actually, the protocol is pretty poorly designed (in more than one way), and one of those shortcomings is it requires a NUL byte to included in the parameter (and figured into the parameter size). If you're just doing one message at a time, you can just use the terminating NUL from the string for the parameter NUL, but if you're doing two things at the same time, you're going to have to explicitly add it in between... so:

```
int c = asprintf(&plymouth_message, 
             "C\x02\x09shutdown%c"
             "M\x02%c%s",
             '\x00', strlen(message) + 1, message);
```

or

```
int c = asprintf(&plymouth_message, 
             "C\x02%c%s%c"
             "M\x02%c%s",
             strlen("shutdown") + 1, "shutdown", '\x00'
             strlen(message) + 1, message);
```

or

```
const char *new_mode = "shutdown";
int c = asprintf(&plymouth_message, 
             "C\x02%c%s%c"
             "M\x02%c%s",
             strlen(new_mode) + 1, new_mode, '\x00'
             strlen(message) + 1, message);
```

You could hide some of the ugliness behind enums and variadic args. it would make it bigger, not sure if that's a good idea or not tbh:

```
typedef enum PlymouthRequest {
    PLYMOUTH_CHANGE_MODE_REQUEST = 'C',
    PLYMOUTH_SHOW_MESSAGE_REQUEST = 'M',
    PLYMOUTH_DONE_WITH_REQUESTS = 0
} PlymouthRequest;

void send_plymouth_requests(int fd, ...) {
    PlymouthRequest request;
    const size_t max_request_argument_length = UCHAR_MAX - 1;
    const size_t max_request_length = strlen ("X\002\255") + 255;
    va_list initial_pass_args, main_pass_args;
    int number_of_requests = 0;
    _cleanup_free_ char *buffer = NULL;
    size_t c = 0;
    int r;

    va_start(main_pass_args, fd);
    va_copy(initial_pass_args, main_pass_args);
    while(va_arg(initial_pass_args, request_type)  != PLYMOUTH_DONE_WITH_REQUESTS) {
        char *argument = va_arg(initial_pass_args, char *);
        if (strlen (arguent) > max_request_argument_length)
            fail...;
        number_of_requests++;
    }
    va_end(count_args);

    buffer = (char*) calloc(number_of_requests, max_request_length);
    if (buffer == NULL)
        fail...;

    while((request = va_arg(main_pass_args, request_type)) != PLYMOUTH_DONE_WITH_REQUESTS) {
        const char *argument = va_arg(main_pass_args, char *);

        r = snprintf(buffer + total_bytes_written, max_request_length, "%c%c%s",
                              (char) request, (char) strlen(msg), msg);
        if (r != strlen(msg) + 2)
            fail...

        c += r + 1 /* NUL byte between requests */;
    }

    va_end(main_pass_args);

    return loop_write(fd, buffer, c, /* do_pool= */ false);
}
```

But if if you did that then you could just:

```
r = send_plymouth_requests (PLYMOUTH_CHANGE_MODE_REQUEST, "shutdown",
                                                     PLYMOUTH_SHOW_MESSAGE_REQUEST, message,
                                                     PLYMOUTH_DONE_WITH_REQUESTS);
```

from the caller... Probably not worth the added complexity (and I just typed it right into this comment box without testing or anything so it probably has bugs), but I'm just throwing it out there as food for thought)

Reply:
Wow, thank you so much @halfline . Your suggestion that I test locally has made a huge difference in my learning.
I'll also try these suggestions you gave and choose.

Thank you so much sir.